### ISSUE TITLE
Discussion : compile time speed

### ISSUE THREAD

**Issue by pfeatherstone** [davisking/dlib#2638]
This isn't a bug report, more of a discussion. (Dlib could do with a Discussion tab)

I've noticed that dlib uses template recursion a lot. I think it's a well known thing that template recursion is slow. @krzysztof-jusiak did a really good [talk](https://www.youtube.com/watch?v=LfOh0DwTP00) on accessing the Nth element of parameter packs. No surprise, template recursion is the slowest. Where this is particularly relevant is the dnn stuff.

Sometimes this can't be avoided, but other times you can replace it with fold expressions, or if using C++11,14, use the parameter pack expansion trick into a dummy initializer list. There are other more performant compile-time algorithms.

With regards to dnn specifically, I think the template model could be refactored with something that doesn't use nested templates so heavily. [In this Cppcast talk](https://cppcast.com/edouard-alligand/), the author of [brigand](https://github.com/edouarda/brigand) mentions that nested templates will be quite slow and lead to very long symbol names which will cripple the compiler. I feel like this is why yolo models are incredibly slow to compile in dlib as they consist of over 400 layers.

Why not replace the current model:
```
using model = layer<layer<layer<layer<layer<>>>>>;
```
with
```
using model = type_list<layer,layer,layer,layer,layer>;
```
then use something like the brigand library to access the Nth layer either using compile time integers or tag types.

So I think there are a few things that could be done which would really improve compiler time performance in dlib. And who knows, maybe it would improve runtime speeds too.

**Comment by pfeatherstone** [comment]
I see no reason why we couldn't have models described like so:
```
using convblock = type_list<con_<...>,bn_<...>,leaky_relu_>;
using model     = type_list<input_layer,convblock,convblock,convblock,loss_layer>;
```
use visitors to set parameters, propagate inputs, back-propagate gradients, etc.

You could access the Nth type by using something like `brigand::at_c<model,N>`

**Comment by davisking** [comment]
Yeah that would probably build a lot faster.  I'm definitely aware of the build times that come from these huge nested templates :|  It wasn't really on my radar in 2015 that people would be making such deep networks. But here we are and the template depth is really excessive now.

**Comment by arrufat** [comment]
> This isn't a bug report, more of a discussion. (Dlib could do with a Discussion tab)

Yes, I'm also in for enabling discussions in here!

**Comment by davisking** [comment]
> > This isn't a bug report, more of a discussion. (Dlib could do with a Discussion tab)
> 
> Yes, I'm also in for enabling discussions in here!

Yeah, just turned that on :D

**Comment by pfeatherstone** [comment]
It could be really cool with concepts at some point

**Comment by davisking** [comment]
> Might give the alternative template model a go at some point. Maybe it could go in a dlib::dnn2 namespace ? I'm hoping it's not too much work.

Yeah that would be cool.

**Comment by VisionEp1** [comment]
yes i am also very much on board with this.

i also did some inspection about what takes so long:
(based on https://devblogs.microsoft.com/cppblog/profiling-template-metaprograms-with-cpp-build-insights/)

I used the yolov5 networks as an example to monitor with vcperf and the windows build insights.
I think there might be also a bug with he repeat layer or the way the impl::layer_helper instantiates templates to perform certain checks. 

 
![image](https://user-images.githubusercontent.com/19327424/185064820-10d66cf2-c4dc-42f5-bf18-e19f115bc62a.png)

as you can see we have over 82k times impl::layer_helper gets called.

So my guess (and I am definitely not a template metaprogramming expert) is that there are 2 discussions to be made:

1. should we work with type lists as @pfeatherstone said
2.  can we improve the current definition in the core.h to be more efficient

**Comment by pfeatherstone** [comment]
What an excellent tool!

**Comment by arrufat** [comment]
Nice, it looks sensible to me.
The only drawback is that I was becoming quite good at defining networks with templates, haha.

**Comment by arrufat** [comment]
Ah, also, I think it would be nice to make the layer constructors take structs, instead, so there should be a
```c++
struct con_options
{
    long num_filters;
    int nr;
    int nc;
    int stride_x;
    int stride_y;
    // ...
};
```

So you coud do something like:
```c++
auto net = module(loss(), con({.num_filters = 32, .nr = 3, .nc = 3}), /* ... */);
```

I don't know, it seems sensible, so we can add more parameters, such as dilation or groups, without changing the signature later on... But maybe there's a reason not to do this.

Finally, how about using the newly created discussions to talk about this? :)

**Comment by pfeatherstone** [comment]
A repeated layer will be used like so:
```
auto model = module(input(), repeat(con(), 4), loss());
```

**Comment by pfeatherstone** [comment]
For reusable blocks, i expect to be able to do this:
```
auto conblock = module(con(), bn(), relu());
auto model = module(input(), repeat(conblock, 4), loss());
```
Something like that. It's definitely doable.

**Comment by pfeatherstone** [comment]
I need to try it out tonight. I've been doing just the type_list approach lately. But then I thought, lets go further and just have a list of lambdas then the model is pretty much just a `std::vector` when constructed.

**Comment by pfeatherstone** [comment]
I like 
```
con({.num_filters = 32, .nr = 3, .nc = 3})
```
a lot. Looks like python almost

**Comment by pfeatherstone** [comment]
Yes we should move to discussion.

**Comment by arrufat** [comment]
I am guessing everything else would work the same, including tags?

**Comment by pfeatherstone** [comment]
The main reason for doing this is that compilers are insanely good at compiling lambdas, it's ridiculous. So using them would be a win. Then having the lambdas collapse to something that wraps `std::vector` when invoked means the symbol names for the models becomes super short and always the same, no matter how complicated your model is.

**Comment by pfeatherstone** [comment]
So hopefully it would be possible to have something like:
```
auto convblock = module(tag(0), con(), bn(), relu(), add(0));
```.
something like that

**Comment by pfeatherstone** [comment]
I think layers that don't require any parameters will just have globally defined lambdas. So convblock would likely be:
```
auto convblock = module(tag(0), con(32,3,3), bn(32), relu, add(0));
```
for example

**Comment by pfeatherstone** [comment]
then to parameterise it you would have:
```
auto convblock = [](size_t outc) {
     return module(tag(0), con(outc, 3, 3), bn(outc), relu, add(0));
};
```
Something like that

**Comment by pfeatherstone** [comment]
Continued here https://github.com/davisking/dlib/discussions/2663
